### PR TITLE
Variables by ID

### PR DESCRIPTION
**Pull request by rachel-fenichel**
## The basics

- [x] I branched from develop
- [x] My pull request is against develop
- [x] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)

### Resolves

https://github.com/google/blockly/issues/1498
https://github.com/google/blockly/issues/1490
https://github.com/google/blockly/issues/1199

### Proposed Changes

- Refer to all variables by ID rather than name.
- Update variable fields: setValue/getValue operate on ID, not variable name.
- All variable fields refer to instances of `Blockly.VariableModel`.
- There is a new "potential variable map" that contains variables that have been referenced in the flyout but not explicitly created (e.g. default values of variable fields).
- Variable Models are created and deleted at appropriate times.
- Undo and redo work.
- Procedures store references to variable models, not just variable names.
- Procedures still only work on variables with the '' type.
- Generators work.  A generator's Blockly.Names object now needs to know about the workspace's variable map.
- Introduce the concept of a "developer variable".  These variables can be referenced in generated code, and will be declared by the generator if a referencing block is used, but are not visible to the user.
  - They won't yet work when used in functions.

### Reason for Changes

Lots of weirdness with variables and events, in Blockly and especially in Scratch Blocks.

### Test Coverage

Tested on:
* Desktop Chrome
* Desktop Firefox

Generator tests work (in Firefox) and JSUnit tests pass.

### Additional Information

I would start with the changes in:
- `core/workspace.js`
- `core/xml.js`
- `core/block.js`
- `core/field_variables.js`
- `core/variable_map.js`

A lot of code moved from `workspace.js` to `variable_map.js`.  `Workspace.js` now has functions that just pass through directly to `variable_map.js`.

`core/variables.js` has some utility functions for dealing with potential and real variables.

Once that's understood, I would move on to the procedure block definitions.

Next I would look at the changes to `core/names.js` and the related changes to the generators and unit tests generator files.

Finally, there area lots of changes in the jsunit tests, since this was a pretty big API change.

### CODE REVIEW

File: core/variable_map.js [hunk 1]
```diff
@@ -164,17 +211,82 @@ Blockly.VariableMap.prototype.deleteVariable = function(variable) {
 };
 
 /**
- * Find the variable by the given name and return it. Return null if it is not
- *     found.
+ * Delete a variables by the passed in ID and all of its uses from this
+ * workspace. May prompt the user for confirmation.
+ * @param {string} id ID of variable to delete.
+ */
+Blockly.VariableMap.prototype.deleteVariableById = function(id) {
```
Comment (AnmAtAnm) [inline core/variable_map.js]: opt_callback with success/cancel

File: blocks/procedures.js [hunk 2]
```diff
@@ -59,6 +59,7 @@ Blockly.Blocks['procedures_defnoreturn'] = {
     this.setTooltip(Blockly.Msg.PROCEDURES_DEFNORETURN_TOOLTIP);
     this.setHelpUrl(Blockly.Msg.PROCEDURES_DEFNORETURN_HELPURL);
     this.arguments_ = [];
+    this.argumentVarModels_ = [];
```
Comment (AnmAtAnm) [inline blocks/procedures.js]: Since argument names can be derived from the argumentVarModels_, is there a point to keep parallel lists?

File: core/block.js [hunk 3]
```diff
@@ -258,6 +271,23 @@ Blockly.Block.prototype.dispose = function(healStack) {
   }
 };
 
+/**
+ * Call initModel on all fields on the block.
+ * May be called more than once.
+ * Either initModel or initSvg must be called after creating a block and before
+ * first interaction with it.
```
Comment (AnmAtAnm) [inline core/block.js]: Nit: ...before _the_ first...
What defines an interaction? Strictly UI, or are some API calls also prohibited before calling init?

File: core/names.js [hunk 4]
```diff
@@ -72,9 +115,18 @@ Blockly.Names.prototype.reset = function() {
  * @return {string} An entity name legal for the exported language.
```
Comment (AnmAtAnm) [inline core/names.js]: "A legal entity name for the exported language." or "An entity name, legal in the exported language."

(unrelated, but...  :/  )

File: core/workspace.js [hunk 5]
```diff
@@ -84,6 +84,18 @@ Blockly.Workspace = function(opt_options) {
...
+   * Blocks in the flyout can refer to variables that don't exist in the
+   * workspace.  For instance, the "get item in list" block refers to an "item"
+   * variable regardless of whether the variable has been created yet.
+   * A FieldVariable must always refer to a Blockly.VariableModel.  We reconcile
+   * these by tracking "potential" variables in the flyout.  These variables
+   * become real when references to them are dragged into the main workspace.
+   * @type {!Blockly.VariableMap}
+   * @private
+   */
+  this.potentialVariableMap_ = new Blockly.VariableMap(this);
```
Comment (AnmAtAnm) [inline core/workspace.js]: This doesn't feel right. It implies the flyout workspaces (and mutator workspaces?) also have their own independent potential variables maps.

This might be something that should be null by default, and set during inject(..), assigning the flyout workspace's variable map as the potential variable map of the main workspace.

**Comment by CleoQc** [comment]
_Introduce the concept of a "developer variable". These variables can be referenced in generated code, and will be declared by the generator if a referencing block is used, but are not visible to the user.
    They won't yet work when used in functions._

Thanks. That seems to be the answer to the question I posted on GoogleGroups.

**Comment by rachel-fenichel** [comment]
@CleoQc yes, it is!  I was waiting going to respond today and point at this PR.

This almost covers your use case--I think the developer variables don't actually work inside functions yet, but I know how to add that.  I need to add developer variables here: https://github.com/google/blockly/blob/master/generators/python/procedures.js#L39

I'm hoping to finish this PR by the end of the week, then follow up with that simple change.  If you're able to hold off for a week, it should be a simple fix on your side.

File: core/variables.js [hunk 6]
```diff
@@ -338,12 +362,122 @@ Blockly.Variables.promptName = function(promptText, defaultText, callback) {
...
+ * @param {string} id The ID to use to look up the variable, or null.
+ * @param {string=} opt_name The string to use to look up the variable.  Only
+ *     used if lookup by ID fails.
+ * @param {string=} opt_type The type to use to look up the variable.  Only used
+ *     if lookup by ID fails.
+ * @return {?Blockly.VariableModel} The variable corresponding to the given ID
+ *     or name + type combination, or null if not found.
+ * @private
+ */
+Blockly.Variables.getVariable = function(workspace, id, opt_name, opt_type) {
```
[the remark this answers is not in the document]
Comment (rachel-fenichel) [reply core/variables.js]: "Further, it the case for querying potential variables needs to be a option flag or another function."

Can't parse.  What does this mean?

File: blocks/procedures.js (hunk 2 above)
Comment (rachel-fenichel) [reply blocks/procedures.js]: I think I can get rid of the arguments_ list.  I didn't yet because it would mean reaching into more code and expanding the scope of this (already large) PR.  I'll make an issue to clean that up.

Comment (rachel-fenichel) [reply blocks/procedures.js]: https://github.com/google/blockly/issues/1532

File: core/names.js (hunk 4 above)
Comment (rachel-fenichel) [reply core/names.js]: I think it's technically correct english!  But yeah, I can change that.

File: core/block.js (hunk 3 above)
Comment (rachel-fenichel) [reply core/block.js]: Updated comment: "Interactions include UI actions (e.g. clicking and dragging) and firing events (e.g. create, delete, and change)."

File: core/variables.js [hunk 7]
```diff
@@ -338,12 +362,122 @@ Blockly.Variables.promptName = function(promptText, defaultText, callback) {
...
+  var potentialVariableMap = workspace.getPotentialVariableMap();
+  // Try to just get the variable, by ID if possible.
+  if (id) {
+    // Look in the real variable map before checking the potential variable map.
+    var variable = workspace.getVariableById(id);
+    if (!variable && potentialVariableMap) {
+      variable = potentialVariableMap.getVariableById(id);
+    }
+  } else if (opt_name && (opt_type != undefined)){
```
[the remark this answers is not in the document]
Comment (rachel-fenichel) [reply core/variables.js]: Correct.  You can uniquely identify a variable by ID or (name + type).

File: core/block.js [hunk 8]
```diff
@@ -686,18 +716,57 @@ Blockly.Block.prototype.getVars = function() {
...
+          variable.getId() == field.getValue()) {
+        field.setText(variable.name);
+      }
+    }
+  }
+};
+
 /**
  * Notification that a variable is renaming.
  * If the name matches one of this block's variables, rename it.
```
Comment (kchadha) [inline core/block.js]: This comment needs to be updated to reflect that the variable id is being checked, not variable name.

File: core/field_variable.js [hunk 9]
```diff
@@ -68,23 +77,40 @@ Blockly.FieldVariable.prototype.init = function() {
...
-  }
-  // If the selected variable doesn't exist yet, create it.
-  // For instance, some blocks in the toolbox have variable dropdowns filled
-  // in by default.
-  if (!this.sourceBlock_.isInFlyout) {
-    this.sourceBlock_.workspace.createVariable(this.getValue());
+  if (this.variable_) {
+    return; // Initialization already happened.
+  }
+  this.workspace_ = this.sourceBlock_.workspace;
```
Comment (picklesrus) [inline core/field_variable.js]: This is slightly different than before and I don't understand why. Mostly I don't understand why the workspace.targetWorkspace is a thing we have at all, so I can't justify to myself why it is okay for it not to be there anymore :)

File: core/field_variable.js [hunk 10]
```diff
@@ -96,39 +122,79 @@ Blockly.FieldVariable.prototype.setSourceBlock = function(block) {
...
  */
 Blockly.FieldVariable.prototype.getValue = function() {
-  return this.getText();
+  return this.variable_ ? this.variable_.getId() : null;
 };
 
 /**
- * Set the variable name.
- * @param {string} value New text.
+ * Get the text from this field.
```
Comment (picklesrus) [inline core/field_variable.js]: Should we leave the part in the comment about this being a variable name as well?

File: core/xml.js [hunk 11]
```diff
@@ -691,13 +728,44 @@ Blockly.Xml.domToBlockHeadless_ = function(xmlBlock, workspace) {
...
+ *     deserialized.
+ * @param {!Element} xml The field tag to decode.
+ * @param {string} text The text content of the XML tag.
+ * @param {!Blockly.FieldVariable} field The field on which the value will be
+ *     set.
+ * @private
+ */
+Blockly.Xml.domToFieldVariable_ = function(workspace, xml, text, field) {
+  var type = xml.getAttribute('variabletype') || '';
+  // TODO (fenichel): Does this need to be explicit or not?
```
Comment (picklesrus) [inline core/xml.js]: is this still a TODO?

File: core/names.js [hunk 12]
```diff
@@ -62,19 +70,63 @@ Blockly.Names = function(reservedWords, opt_variablePrefix) {
...
   this.dbReverse_ = Object.create(null);
+  this.variableMap_ = null;
+};
+
+/**
+ * Set the variable map that maps from variable name to variable object.
+ * @param {!Blockly.VariableMap} map The map to track.
+ * @package
+ */
+Blockly.Names.prototype.setVariableMap = function(map) {
```
Comment (picklesrus) [inline core/names.js]: Should this just be passed in to the constructor of Names?

  At the same time I suggest that, I'm also not sure the Names object should depend on the workspace.

File: generators/php.js [hunk 13]
```diff
@@ -149,14 +149,23 @@ Blockly.PHP.init = function(workspace) {
...
-  var variables = Blockly.Variables.allVariables(workspace);
+  var variables = workspace.getAllVariables();
   for (var i = 0, variable; variable = variables[i]; i++) {
-    varName = variable.name;
-    defvars[i] = Blockly.PHP.variableDB_.getName(varName,
+    defvars[i] = Blockly.PHP.variableDB_.getName(variable.getId(),
         Blockly.Variables.NAME_TYPE) + ';';
   }
+
+  // Add developer variables (not created or named by the user).
```
Comment (picklesrus) [inline generators/php.js]: A lot of the developer variable stuff in generator stuff is pretty repetitive. e.g. this loop is almost the same in all of them. Is there a way to avoid that?

File: core/xml.js (hunk 11 above)
Comment (rachel-fenichel) [reply core/xml.js]: Yes!  I can't convince myself one way or another.  I *think* if I don't make it explicit this way, it just doesn't get serialized if it's empty.

File: tests/jsunit/event_test.js [hunk 14]
```diff
@@ -573,3 +671,75 @@ function helper_addMoveEventParent(events, block, parent) {
...
+
+    // Expect the group ID to be the result of the second call to genUid.
+    assertEquals('2', event0.group);
+
+    // Expect the workspace to have a variable with ID '3'.
+    assertNotNull(workspace.getVariableById('3'));
+    assertEquals('3', event0.varId);
+  } finally {
+    eventTest_tearDownWithMockBlocks();
+    Blockly.Events.fire = savedFireFunc;
```
Comment (AnmAtAnm) [inline tests/jsunit/event_test.js]: Is there not a way to set this before every test function, instead of relying on individual tests to remember to set it it back?

Other test files use a convention of setup/teardown pair of methods that would be appropriate for this.

The whole setup/try/finally/teardown process could be simplified via a wrapper function builder:

```js
var test_func_whatever = buildTest(function() { ... });  // Or arrow func

function buildTest(testImp) {
  return function() {
    setup();
    try { testImpl(); }
    finally { teardown(); }
    // Maybe return any result from testImpl()
  }
}
```


All of this can be a later PR.

File: tests/jsunit/event_test.js [hunk 15]
```diff
@@ -573,3 +671,75 @@ function helper_addMoveEventParent(events, block, parent) {
...
+  temporary_fireEvent.firedEvents_ = [];
+  // Expect three calls to genUid: one to set the block's ID, one for the event
+  // group's id, and one for the variable's ID.
+  setUpMockMethod(mockControl_, Blockly.utils, 'genUid', null, ['1', '2', '3']);
+  try {
+    var block = workspace.newBlock('field_variable_test_block');
+
+    var firedEvents = workspace.undoStack_;
+    // Expect two events: varCreate and block create.
+    assertEquals(2, firedEvents.length);
```
Comment (AnmAtAnm) [inline tests/jsunit/event_test.js]: Prefer to use assert messages (optional leading argument: http://www.jsunit.net/documentation/assertions/) instead of code comments.  Using a assert message will expose the expectation in the test logs upon failure.

File: tests/jsunit/test_utilities.js [hunk 16]
```diff
@@ -89,3 +89,54 @@ function checkVariableValues(container, name, type, id) {
...
 }
+
+/**
+ * Create a test get_var_block.
+ * Will fail if get_var_block isn't defined.
+ * TODO (fenichel): Rename to createMockVarBlock.
+ * @param {!string} variable_id The id of the variable to reference.
+ * @return {!Blockly.Block} The created block.
+ */
+function createMockBlock(variable_id) {
```
Comment (AnmAtAnm) [inline tests/jsunit/test_utilities.js]: Perhaps createMockGetVarBlock(..)
(I know this was just a move and was already in use)

File: core/field_variable.js [hunk 17]
```diff
@@ -96,39 +122,79 @@ Blockly.FieldVariable.prototype.setSourceBlock = function(block) {
...
+  // Type checks!
+  var type = variable.type;
+  if (!this.typeIsAllowed_(type)) {
+    throw new Error('Variable type doesn\'t match this field!  Type was ' +
+        type);
+  }
+  if (this.sourceBlock_ && Blockly.Events.isEnabled()) {
+    var oldValue = this.variable_ ? this.variable_.getId() : null;
+    Blockly.Events.fire(new Blockly.Events.BlockChange(
+        this.sourceBlock_, 'field', this.name, oldValue, variable.getId()));
```
Comment (RoboErikG) [inline core/field_variable.js]: nit: Could use id instead of variable.getId()

File: core/field_variable.js [hunk 18]
```diff
@@ -138,6 +204,8 @@ Blockly.FieldVariable.prototype.setValue = function(value) {
  * @private
  */
 Blockly.FieldVariable.prototype.getVariableTypes_ = function() {
+  // TODO: Why does this happen every time, instead of once when the workspace
```
Comment (RoboErikG) [inline core/field_variable.js]: Because we don't have a way to specify variable types ahead of time, so this list depends on which variables have been created so far. Part of issue #1513

File: core/variable_map.js [hunk 19]
```diff
@@ -164,17 +214,85 @@ Blockly.VariableMap.prototype.deleteVariable = function(variable) {
...
+/**
+ * Deletes a variable and all of its uses from this workspace without asking the
+ * user for confirmation.
+ * @param {!Blockly.VariableModel} variable Variable to delete.
+ * @param {!Array.<!Blockly.Block>} uses An array of uses of the variable.
+ * @private
+ */
+Blockly.VariableMap.prototype.deleteVariableInternal_ = function(variable,
+    uses) {
+  Blockly.Events.setGroup(true);
```
Comment (RoboErikG) [inline core/variable_map.js]: Move the calls to setGroup up to the public method. Internal methods might be chained so they shouldn't change the grouping flag.

File: core/field_variable.js [hunk 20]
```diff
@@ -47,15 +47,24 @@ goog.require('goog.string');
...
-      Blockly.FieldVariable.dropdownCreate, opt_validator);
-  this.setValue(varname || '');
+  // The FieldDropdown constructor would call setValue, which might create a
+  // spurious variable.  Just do the relevant parts of the constructor.
+  this.menuGenerator_ = Blockly.FieldVariable.dropdownCreate;
+  this.size_ = new goog.math.Size(0, Blockly.BlockSvg.MIN_BLOCK_Y);
+  this.setValidator(opt_validator);
+  // TODO (#1499): Add opt_default_type to match default value.  If not set, ''.
+  this.defaultVariableName = (varname || '');
+  this.defaultType_ = '';
```
Comment (kchadha) [inline core/field_variable.js]: I know there's a TODO here to add an opt_default_type, but I think for the time being, `defaultType_` can be smarter in the case where opt_variableTypes is actually provided and non-empty.

e.g. something like:

```
var hasVariableTypes = (opt__variableTypes && opt_variableTypes.length > 0);
this.defaultType_ = hasVariableTypes ? opt_variableTypes[0] : '';
```
OR

```
var hasSingleVariableType = (opt__variableTypes && opt_variableTypes.length == 1);
this.defaultType_ = hasSingleVariableType ? opt_variableTypes[0] : '';
```

Having the default type set correctly is definitely very useful when integrating with scratch-blocks.

File: core/variable_map.js [hunk 21]
```diff
@@ -219,11 +337,21 @@ Blockly.VariableMap.prototype.getVariablesOfType = function(type) {
 };
 
 /**
- * Return all variable types.
+ * Return all variable types.  This list always contains the empty string.
```
Comment (kchadha) [inline core/variable_map.js]: I'm curious about the reason for this change...

File: core/workspace.js [hunk 22]
```diff
@@ -123,18 +135,6 @@ Blockly.Workspace.SCAN_ANGLE = 3;
...
-  }
-  // This is for the (unlikely) case where you have a variable in a block in
-  // an always-open flyout.  It needs to be possible to edit the block in the
-  // flyout, so the contents of the dropdown need to be correct.
-  var variableNames = Blockly.Variables.allUsedVariables(block);
-  for (var i = 0, name; name = variableNames[i]; i++) {
-    if (!this.getVariable(name)) {
-      this.createVariable(name);
-    }
-  }
```
Comment (kchadha) [inline core/workspace.js]: What is the reason for this change?

File: core/variables.js [hunk 23]
```diff
@@ -64,13 +64,13 @@ Blockly.Variables.allUsedVariables = function(root) {
...
     if (blockVariables) {
       for (var y = 0; y < blockVariables.length; y++) {
-        var varName = blockVariables[y];
-        // Variable name may be null if the block is only half-built.
-        if (varName) {
-          variableHash[varName.toLowerCase()] = varName;
+        var variable = blockVariables[y];
+        // Variable ID may be null if the block is only half-built.
+        if (variable.getId()) {
+          variableHash[variable.name.toLowerCase()] = variable.name;
```
Comment (kchadha) [inline core/variables.js]: Does it make sense for this function to return either a list of variable IDs or VariableModels instead of variable names since variables are no longer unique by name?

File: core/field_variable.js (hunk 20 above)
Comment (kchadha) [reply core/field_variable.js]: Oops, just realized we've already had this discussion before. You can ignore this.

File: core/block.js [hunk 24]
```diff
@@ -155,8 +155,21 @@ Blockly.Block = function(workspace, prototypeName, opt_id) {
   // Record initial inline state.
   /** @type {boolean|undefined} */
   this.inputsInlineDefault = this.inputsInline;
+
+  // Fire a create event.
   if (Blockly.Events.isEnabled()) {
-    Blockly.Events.fire(new Blockly.Events.BlockCreate(this));
+    var existingGroup = Blockly.Events.getGroup();
```
[the remark this answers is not in the document]
Comment (rachel-fenichel) [reply core/block.js]: Creation of a single block didn't need to be in a group before variables started to have IDs and events.  

Creating a block may also create a variable, which fires an event.  If it does, they should be grouped together so that undoing block creation also undoes variable creation.

Sometimes we avoid this problem if the secondary events are deterministic (moving blocks around when changing a proecdure, for instance).  The variable creation is deterministic (for the same XML, on the same workspace) but the ID of the new variable is not--it's a UUID.  So the var create event will need to be fired.

The final piece of the puzzle: the variable creation happens at the first "interaction" with the variable.  In this case, that's when the block is serialized for the BlockCreate event.  Most of the time we avoid this problem, because when loading blocks from XML we turn off events until all of the fields on the blocks have been set.

File: core/field_variable.js (hunk 9 above)
Comment (rachel-fenichel) [reply core/field_variable.js]: Mostly this just moved into getOrCreateVariable, and the concept of potential variables has replaced whatever we were doing for flyout variables before.  

A flyout workspace has a target workspace; the main workspace doesn't.  In a flyout workspace we don't want to create a real variable if we don't have to, because it wouldn't go away when the flyout closes.  Previously we were just coming up with a unique name and calling setValue, but not actually creating a variable model.  Now it's required to act on a variable model, so we create one.

File: core/field_variable.js (hunk 17 above)
Comment (rachel-fenichel) [reply core/field_variable.js]: For reasons I don't remember we made the variable ID private and always access it with getIde().  It doesn't actually change anything, but for convention I'm not accessing private variables directly.

File: core/field_variable.js (hunk 18 above)
Comment (rachel-fenichel) [reply core/field_variable.js]: I'll change the TODO to poitn to #1513.

File: core/variable_map.js (hunk 19 above)
Comment (rachel-fenichel) [reply core/variable_map.js]: I'd rather make this check if the group is already set and use that to decide whether to call setGroup(false).  Moving the setGroup calls out of here entirely feels error-prone.

File: core/workspace.js (hunk 22 above)
Comment (rachel-fenichel) [reply core/workspace.js]: For getting rid of this?  Because we're now creating the variable as part of initModel or in XML deserialization, and doing it in all cases instead of just in this special case.  Also because that case is obviously not unlikely if it happens on every single use of scratch-blocks.

File: core/field_variable.js (hunk 17 above)
Comment (RoboErikG) [reply core/field_variable.js]: but you use "id" directly three lines later. The local method scoped one, not the class field.

File: core/variable_map.js (hunk 19 above)
Comment (RoboErikG) [reply core/variable_map.js]: Adding the check like in block creation also works. I'll file a bug that we should do a pass on our public APIs to add setGroup wrappers.

File: core/names.js (hunk 4 above)
Comment (AnmAtAnm) [reply core/names.js]: Agreed, but my brain backfires when trying to read the original form.

File: core/variables.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (AnmAtAnm) [reply core/variables.js]: failing silently seems weird, but okay for a private method.

File: core/field_variable.js (hunk 10 above)
Comment (rachel-fenichel) [reply core/field_variable.js]: Done.

File: core/field_variable.js (hunk 17 above)
Comment (rachel-fenichel) [reply core/field_variable.js]: Done.

File: core/variables.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (rachel-fenichel) [reply core/variables.js]: Updated to throw an error.

File: core/variables.js [hunk 25]
```diff
@@ -338,12 +362,122 @@ Blockly.Variables.promptName = function(promptText, defaultText, callback) {
...
+ * @param {!Blockly.Workspace} workspace The workspace in which to create the
+ * variable.  It may be a flyout workspace or main workspace.
+ * @param {string} id The ID to use to create the variable, or null.
+ * @param {string} name The string to use to create the variable.
+ * @param {string} type The type to use to create the variable.
+ * @return {!Blockly.VariableModel} The variable corresponding to the given ID
+ *     or name + type combination.
+ * @private
+ */
+Blockly.Variables.createVariable_ = function(workspace, id, name, type) {
```
[the remark this answers is not in the document]
Comment (rachel-fenichel) [reply core/variables.js]: Done, and I updated the annotations here and on getOrCreateVariable.

File: core/workspace.js [hunk 26]
```diff
@@ -84,6 +84,18 @@ Blockly.Workspace = function(opt_options) {
    * @private
    */
   this.variableMap_ = new Blockly.VariableMap(this);
+
+  /**
+   * Blocks in the flyout can refer to variables that don't exist in the
+   * workspace.  For instance, the "get item in list" block refers to an "item"
```
[the remark this answers is not in the document]
Comment (rachel-fenichel) [reply core/workspace.js]: Done.

File: core/workspace.js [hunk 27]
```diff
@@ -311,65 +228,11 @@ Blockly.Workspace.prototype.createVariable = function(name, opt_type, opt_id) {
 
 /**
  * Find all the uses of a named variable.
```
[the remark this answers is not in the document]
Comment (rachel-fenichel) [reply core/workspace.js]: Done.

File: core/workspace.js [hunk 28]
```diff
@@ -439,6 +295,35 @@ Blockly.Workspace.prototype.getVariableById = function(id) {
...
+Blockly.Workspace.prototype.getVariablesOfType = function(type) {
+  return this.variableMap_.getVariablesOfType(type);
+};
+
+/**
+ * Return all variable types.
+ * @return {!Array.<string>} List of variable types.
+ */
+Blockly.Workspace.prototype.getVariableTypes = function() {
+  return this.variableMap_.getVariableTypes();
```
[the remark this answers is not in the document]
Comment (rachel-fenichel) [reply core/workspace.js]: I don't know.  I think this is also part of #1513

File: core/xml.js [hunk 29]
```diff
@@ -86,6 +86,34 @@ Blockly.Xml.blockToDomWithXY = function(block, opt_noId) {
   return element;
 };
 
+/**
+ * Encode a variable field as XML.
+ * @param {!Blockly.Field} field The field to encode.
```
[the remark this answers is not in the document]
Comment (rachel-fenichel) [reply core/xml.js]: Yes.  Fixed.

File: core/xml.js [hunk 30]
```diff
@@ -86,6 +86,34 @@ Blockly.Xml.blockToDomWithXY = function(block, opt_noId) {
...
+ * @return {?Element} XML element, or null if the field did not need to be
+ *     serialized.
+ * @private
+ */
+Blockly.Xml.fieldToDomVariable_ = function(field) {
+  var id = field.getValue();
+  // The field had not been initialized fully before being serialized.
```
[the remark this answers is not in the document]
Comment (rachel-fenichel) [reply core/xml.js]: Added to comment:
```
//This can happen if a block is created directly through a call to
// workspace.newBlock instead of from XML.
// The new block will be serialized for the first time when firing a block
// creation event.
```

File: core/xml.js [hunk 31]
```diff
@@ -86,6 +86,34 @@ Blockly.Xml.blockToDomWithXY = function(block, opt_noId) {
...
+  if (id == null) {
+    field.initModel();
+    id = field.getValue();
+  }
+  // Get the variable directly from the field, instead of doing a lookup.  This
+  // will work even if the variable has already been deleted.
+  var variable = field.getVariable();
```
[the remark this answers is not in the document]
Comment (rachel-fenichel) [reply core/xml.js]: Added to the comment:
```
  // This can happen
  // because the flyout defers deleting blocks until the next time the flyout is
  // opened.
```

File: generators/php.js (hunk 13 above)
Comment (rachel-fenichel) [reply generators/php.js]: Probably, but I'm punting on that API cleanup for this PR.  Here's a bug for it: https://github.com/google/blockly/issues/1535

File: core/workspace.js (hunk 5 above)
Comment (rachel-fenichel) [reply core/workspace.js]: Fixed.

**Comment by rachel-fenichel** [comment]
Travis is still failing due to an outage.  I've confirmed that the tests pass locally.

**Comment by CleoQc** [comment]
I'm really happy to see this get merged.   I do have a couple of questions that are unrelated to the code per se.
When is the hoped-for target date to see this in the main branch?
Is there a write-up or a simple example on how to use developer variables?

Thanks a lot for all the work going into this!

**Comment by rachel-fenichel** [comment]
Target date to get this into master: probably 1-2 months out.  This is a big enough change that I expect plenty of small followup-up changes or bugs.

Write-up or simple example for developer variables: not yet.  Let's move the discussion over to #1535 and I'll try to answer any questions you have.  That will also help me sort out the API.
